File: ver/Python/t-NPLQCD-File-Format-SDB.py
#!/usr/bin/env python

# === Import ...
import sys
import hashlib
import time
import json
#///

from NPLQCD.File.Format.SDB import sdb_dump, digest, canonical_dump, c_qdpxx_sdb

#
# Taken from NPLQCD/File/Format/SDB/SDB.py
#

if __name__ == '__main__': # ===

 verbose, validate_only, *files = sys.argv[1:]

 # === command line args ...

 opt = \
 {
   'redirect': True
 , 'timeout' :
   {
     'skip':   0.001
   , 'step': 100
   }
 }

 def as_bool(s): return { 'True': True, 'False': False }[s]

 verbose       = as_bool(verbose)
 validate_only = as_bool(validate_only)

#///

 hash=hashlib.md5

 time_ = []

 time_.append( time.perf_counter() )

 ret = sdb_dump(files,opt=opt,verbose=verbose,validate_only=validate_only,hash=hash)

 time_.append( time.perf_counter() )

 print('{} VALIDITY'.format( digest(hash,ret['valid']) ))
  
 if not validate_only:
  print('{} DATA     // SDB::sdb_dump     elapsed {:8.3f}'.format( digest(hash,canonical_dump(ret['data'],'json')), time_[-1] - time_[-2] ) )

 valid_ = \
 {
  each['file']: each['valid']
   for each in json.loads(ret['valid'])
 }

 time_.append( time.perf_counter() )

 ret_ = \
 [
  { file: c_qdpxx_sdb.dump(file) } 
   for file in files
    if valid_[file]
 ]

 time_.append( time.perf_counter() )

 if not validate_only:
  print('{} DATA     // c_qdpxx_sdb::dump elapsed {:8.3f}'.format( digest(hash,canonical_dump(ret_,'json')), time_[-1] - time_[-2] ) )


#///

File: ver/Perl/t-NPLQCD-File-Format-SDB.pl
#!/usr/bin/env perl

# === Use ...

use lib '.' ;
use strict ;

$::False = \0 ;
$::True  = \1 ;

use JSON::XS ;
use Data::Dump qw/dump/ ;
use Digest::MD5 ;
use Time::HiRes qw/clock_gettime/ ;

#///

use Package::Alias
  'SDB'         => 'NPLQCD::File::Format::SDB'

#
# Already aliased from above.
#
# , 'c_qdpxx_sdb' => 'NPLQCD::File::Format::SDB::XS'
#
;

use SDB qw/sdb_dump digest canonical_dump/ ;

__PACKAGE__->__main__ ;

#
# Taken from NPLQCD::File::Format::SDB.pm
#

sub __main__ # ===
{
my ($verbose,$validate_only,@file) = @ARGV ;

# === command line args ...

my $opt =
 {
   'redirect' => 0
 , 'timeout'  =>
   {
     'skip' =>   0.001
   , 'step' => 100
   }
 } ;

my $as_bool = sub { { 'True' => $::True, 'False' => $::False }->{(shift)} ; } ;

 $verbose       = $as_bool->($verbose) ;
 $validate_only = $as_bool->($validate_only) ;

#///

my $hash = \&Digest::MD5::md5 ;

my @time ;

push @time, clock_gettime ;

#
# Default options is equivalent to
#
#  my $ret = SDB::sdb_dump( \@file ) ;
#
my $ret = sdb_dump( \@file,'opt' => $opt, 'verbose' => $verbose, 'validate_only' => $validate_only, 'hash' => $hash ) ;

push @time, clock_gettime ;

 printf("%s VALIDITY\n", digest($hash,$ret->{'valid'}) ) ;

 if( not $$validate_only )
 {
  printf("%s DATA     // SDB::sdb_dump     elapsed %8.3f (s)\n", digest($hash,canonical_dump($ret->{'data'},"json")), $time[-1] - $time[-2] ) ;
 }

my %valid =
 map
 {
  $_->{'file'} => $_->{'valid'}
 } @{JSON::XS->new->decode($ret->{'valid'})}
;

#
# If you are certain that the SDB files are free from errors,
# you may access them directly.
#

push @time, clock_gettime ;

my @ret =
 map
 {
  my $hash = { $_ => c_qdpxx_sdb::dump($_) } ;

  $hash

 }
  grep { $valid{$_} }
   @file 
;

push @time, clock_gettime ;

 if( not $$validate_only )
 {
  printf("%s DATA     // c_qdpxx_sdb::dump elapsed %8.3f (s)\n", digest($hash,canonical_dump(\@ret,"json")), $time[-1] - $time[-2] ) ;
 }

}

#///

